shared/src/main/scala/mlscript/codegen/typescript/TsTypegen.scala

```diff
@@ -11,37 +11,85 @@ import mlscript.TypeDef
-  private val typegenCode: MutListBuffer[SourceCode] = MutListBuffer.empty;
-
-  /** Return complete generated typegen code
+  // store converted mlscript type definitions and terms created in mlscript
+  // use typeScope and termScope to generate non-conflicting names for both
+  private val typeScope: Scope = Scope()
+  private val termScope: Scope = Scope()
+  private val typegenCode: ListBuffer[SourceCode] = ListBuffer.empty;
+
+  /** Return complete typegen code for current typing unit
     *
     * @return
     *   SourceCode
     */
   def toSourceCode(): SourceCode = {
-    for (code <- typegenCode) {
-      println(code.toString())
-    }
     SourceCode("// start ts") +=: typegenCode
     typegenCode += SourceCode("// end ts")
     SourceCode.bulkConcat(typegenCode)
   }
 
-  /** Converts a definition to its typescript type declarations, including any new type aliases created in order to
-    * represent it's type.
+  /** Context for converting a statement or type definition
+    * 
+    *
+    * @param existingTypeVars
+    *   type variables defined for the current type
+    * @param typeVarMapping
+    *   mapping from type variables to their clash-free parameter names
+    * @param termScope
+    *   scope for argument names
+    * @param typeScope
+    *   scope for type parameter names
+    */
+  protected case class TypegenContext(
+      val existingTypeVars: Map[TypeVar, String],
+      // Mapping type variables existing and new ones
+      // to clash-free type names generated for current typing unit
+      val typeVarMapping: MutMap[TypeVar, String],
+      // a scope for generating unique argument names
+      val termScope: Scope,
+      // a scope for generating unique type parameter names
+      val typeScope: Scope
+  ) {
+    /** Try to find name for type variable in mapping. If it
+      * doesn't exist check existing type variable names mapping
+      * 
+      * @param t TypeVar
+      * @return
+      *   Left[String, None] is from type var mapping
+      *   Right[None, String] is from existing type variable names
+      */
+    def findTypeVarName(t: TypeVar): Either[String, String] = {
+      this.typeVarMapping.get(t).map(Left(_)).getOrElse({
+        this.existingTypeVars
+          .get(t)
+          .map(Right(_))
+          .getOrElse(
+            throw IllFormedTsTypeError(s"Did not find mapping for type variable $t. Unable to generated ts type.")
+          )
+      })
+    }
+  }
+
+  object TypegenContext {
+    def apply(mlType: Type): TypegenContext = {
+      val existingTypeVars = ShowCtx.mk(mlType :: Nil, "").vs;
+      // initialize local term scope with global term scope as parent
+      // this will reduce cases where global names are similar to function
+      // argument names. For e.g. the below case will be avoided
+      // export const arg0: int
+      // export const f: (arg0: int) => int
+      TypegenContext(existingTypeVars, MutMap.empty, Scope(Seq.empty, termScope), Scope())
```

Comment:
Shouldn't the type scope also initialize its parent to the outer type scope?

Reply:
Hmmm yes type aliases could clash with type parameter names. I'll add the parent scope.